wen-restart/src/wen_restart.rs

```diff
@@ -341,14 +375,142 @@ pub(crate) fn find_heaviest_fork(
+        let old_root_bank = my_bank_forks.root_bank();
+        if !old_root_bank
+            .hard_forks()
+            .iter()
+            .any(|(slot, _)| slot == &new_root_slot)
+        {
+            old_root_bank.register_hard_fork(new_root_slot);
+        }
+        // new_root_slot is guaranteed to have a bank in bank_forks, it's checked in
+        // find_bankhash_of_heaviest_fork().
+        match my_bank_forks.get(new_root_slot) {
+            Some(bank) => new_root_bank = bank.clone(),
+            None => {
+                return Err(WenRestartError::BlockNotFound(new_root_slot).into());
+            }
+        }
+        let mut banks = vec![&new_root_bank];
+        let parents = new_root_bank.parents();
+        banks.extend(parents.iter());
+
+        let _ = my_bank_forks.send_eah_request_if_needed(
+            new_root_slot,
+            &banks,
+            accounts_background_request_sender,
+        )?;
+    }
+    // There can't be more than one EAH calculation in progress. If new_root is generated
+    // within the EAH window (1/4 epoch to 3/4 epoch), the following function will wait for
+    // EAH calculation to finish. So if we trigger another EAH when generating snapshots
+    // we won't hit a panic.
+    let _ = new_root_bank.get_epoch_accounts_hash_to_serialize();
+    // Even though generating incremental snapshot is faster, it involves finding a full
+    // snapshot to use as base, so the logic is more complicated. For now we always generate
+    // a full snapshot, even if one already exists, because after adding a hard fork the
+    // bankhash will change.
+    let mut directory = &snapshot_config.full_snapshot_archives_dir;
+    let Some(full_snapshot_slot) = get_highest_full_snapshot_archive_slot(directory) else {
+        return Err(WenRestartError::MissingFullSnapshot(
+            snapshot_config
+                .full_snapshot_archives_dir
+                .to_string_lossy()
+                .to_string(),
+        )
+        .into());
+    };
+    // In very rare cases it's possible that the local root is not on the heaviest fork, so the
+    // validator generated snapshot for slots > local root. In this case, we don't proceed with
+    // wen_restart, since this needs human inspection.
+    // In even rarer cases, the selected slot might be the last full snapshot slot. We could
+    // just re-generate a new snapshot to make sure the snapshot is up to date after hard fork,
+    // but for now we just return an error to keep the code simple.
+    if full_snapshot_slot >= new_root_slot {
+        return Err(WenRestartError::FutureSnapshotExists(
+            new_root_slot,
+            full_snapshot_slot,
+            directory.to_string_lossy().to_string(),
+        )
+        .into());
+    }
+    directory = &snapshot_config.incremental_snapshot_archives_dir;
+    if let Some(slot) = get_highest_incremental_snapshot_archive_slot(directory, full_snapshot_slot)
```

Comment:
nit: rename to `incremental_snapshot_slot` for consistency

Reply:
Done.